### PR TITLE
Fix Nested Metrics Handling in CompileMetrics

### PR DESCRIPTION
**Pull request by Dimios45**
This PR addresses the issue reported in [Keras Issue #21700](https://github.com/keras-team/keras/issues/21700), where nested output structures and corresponding losses are supported, but nested metrics are not. This caused failures when attempting to compile models with nested metrics configurations.

### Changes
- [x] Implemented recursive handling of nested metric configurations within `CompileMetrics`.
- [x] Ensured compatibility with both nested and flat metric structures.
- [x] Updated internal methods to process nested metrics appropriately.

### Testing
- [x] Added tests to verify correct handling of nested metrics during model compilation and training.
- [x] Tests cover nested dictionaries and lists to ensure the recursive metric construction behaves as expected and raises appropriate errors for invalid configurations.

### Example Usage:

```python
model.compile(
    optimizer='adam',
    loss={'a': 'mse', 'b': {'c': 'mse', 'd': 'mse'}},
    metrics={'a': ['mae'], 'b': {'c': 'mse', 'd': 'mae'}}

### CODE REVIEW

File: keras/src/trainers/compile_utils.py [hunk 1]
```diff
@@ -175,6 +175,40 @@ def variables(self):
...
+            return False
+        if isinstance(obj, dict):
+            for value in obj.values():
+                if isinstance(value, (dict, list)):
+                    return True
+        elif isinstance(obj, list):
+            for item in obj:
+                if isinstance(item, (dict, list)):
+                    return True
+        return False
```
Comment (hertschuh) [inline keras/src/trainers/compile_utils.py:209]: Why not use `tree.is_nested`?
